File: ruby/lib/smith/client/command.rb
#  File: command.rb
#  Base class for commands
#
#  This file is part of the Ember Ruby Gem.
#
#  Copyright 2015 Autodesk, Inc. <http://ember.autodesk.com/>
#  
#  Authors:
#  Jason Lefley
#
#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#
#  THIS PROGRAM IS DISTRIBUTED IN THE HOPE THAT IT WILL BE USEFUL,
#  BUT WITHOUT ANY WARRANTY; WITHOUT EVEN THE IMPLIED WARRANTY OF
#  MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.  SEE THE
#  GNU GENERAL PUBLIC LICENSE FOR MORE DETAILS.
#
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <http://www.gnu.org/licenses/>.

require 'smith/client/url_helper'

module Smith
  module Client
    class Command

      COMPLETED_ACK = 'completed'
      RECEIVED_ACK = 'received'
      FAILED_ACK = 'failed'

      include URLHelper
      include PayloadHelper
      
      def initialize(state, http_client, payload)
        @state, @http_client, @payload = state, http_client, payload
      end

      private 

      # Send an command acknowledgement post request state is the stage of the command acknowledgment
      # If only the state is specified, the message is nil
      # If a string is specified as the second argument, it is formatted as a log message using any additional arguments
      # If something other than a string is specified, it is used directly
      def acknowledge_command(state, *args)
        m = message(*args)
        request = @http_client.post(acknowledge_endpoint(@payload), command_payload(@payload.command, state, m, Printer.get_status))
        request.callback { Client.log_debug(LogMessages::ACKNOWLEDGE_COMMAND, @payload.command, @payload.task_id, state, m) }
      end

      # Get the message portion of the acknowledgment according to the specified arguments
      def message(*args)
        return                          if args.empty?
        return LogMessage.format(*args) if args.first.is_a?(String)
        return args.first
      end

    end
  end
end

File: AVR/MotorController/Command.cpp
//  File: Command.cpp
//  Encapsulates and provides access to a command's register, action, and parameter
//
//  This file is part of the Ember Motor Controller firmware.
//
//  Copyright 2015 Autodesk, Inc. <http://ember.autodesk.com/>
//
//  Authors:
//  Jason Lefley
//
//  This program is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, either version 2 of the License, or
//  (at your option) any later version.
//
//  THIS PROGRAM IS DISTRIBUTED IN THE HOPE THAT IT WILL BE USEFUL,
//  BUT WITHOUT ANY WARRANTY; WITHOUT EVEN THE IMPLIED WARRANTY OF
//  MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.  SEE THE
//  GNU GENERAL PUBLIC LICENSE FOR MORE DETAILS.
//
//  You should have received a copy of the GNU General Public License
//  along with this program.  If not, see <http://www.gnu.org/licenses/>.

#include "Command.h"


Command::Command() :
    mRegister(0),
    mAction(0),
    mParameter(0)
{
}

Command::~Command()
{
}

// Set the command's register value
// value The register value
void Command::SetRegister(unsigned char value)
{
    mRegister = value;
}

// Return the command's register value
unsigned char Command::Register() const
{
    return mRegister;
}

// Set the command's action value
// value The action value
void Command::SetAction(unsigned char value)
{
    mAction = value;
}

// Return the command's action value
unsigned char Command::Action() const
{
    return mAction;
}

// Sets the command's parameter by assembling the comprising bytes
// buffer A byte length of 4 containing the parameter bytes (LSB first)
void Command::SetParameterBytes(const unsigned char* buffer)
{
    int32_t ones = 0xFFFFFFFF;

    mParameter = (buffer[0] & ones) | (buffer[1] & ones) << 8 |
        (buffer[2] & ones) << 16 | (buffer[3] & ones) << 24;
}

// Return the command's parameter value
int32_t Command::Parameter() const
{
    return mParameter;
}

